include/circt/Dialect/FIRRTL/FIRRTLModuleContext.h

```diff
@@ -0,0 +1,173 @@
+  // nodes which are obviously redundant.  Instead of literally producing them
+  // in the parser, do an implicit CSE to reduce parse time and silliness in the
+  // resulting IR.
+  llvm::DenseMap<std::pair<Attribute, Type>, Value> constantCache;
+
+  //===--------------------------------------------------------------------===//
+  // SubaccessCache
+
+  /// This returns a reference with the assumption that the caller will fill in
+  /// the cached value. We keep track of inserted subaccesses so that we can
+  /// remove them when we exit a scope.
+  Value &getCachedSubaccess(Value value, unsigned index);
+
+  //===--------------------------------------------------------------------===//
+  // SymbolTable
+
+  /// Add a symbol entry with the specified name, returning failure if the name
+  /// is already defined.
+  ///
+  /// When 'insertNameIntoGlobalScope' is true, we don't allow the name to be
+  /// popped.  This is a workaround for (firrtl scala bug) that should
+  /// eventually be fixed.
+  ParseResult addSymbolEntry(StringRef name, SymbolValueEntry entry,
+                             llvm::SMLoc loc,
+                             bool insertNameIntoGlobalScope = false);
+
+  /// Look up the specified name, emitting an error and returning null if the
+  /// name is unknown.
+  ParseResult addSymbolEntry(StringRef name, Value value, llvm::SMLoc loc,
+                             bool insertNameIntoGlobalScope = false);
+
+  /// Resolved a symbol table entry to a value.  Emission of error is optional.
+  ParseResult resolveSymbolEntry(Value &result, SymbolValueEntry &entry,
+                                 llvm::SMLoc loc, bool fatal = true);
+
+  /// Resolved a symbol table entry if it is an expanded bundle e.g. from an
+  /// instance.  Emission of error is optional.
+  ParseResult resolveSymbolEntry(Value &result, SymbolValueEntry &entry,
+                                 StringRef field, llvm::SMLoc loc);
+
+  /// Look up the specified name, emitting an error and returning failure if the
+  /// name is unknown.
+  ParseResult lookupSymbolEntry(SymbolValueEntry &result, StringRef name,
+                                llvm::SMLoc loc);
+
+  UnbundledValueEntry &getUnbundledEntry(unsigned index);
+
+  /// This contains one entry for each value in FIRRTL that is represented as a
+  /// bundle type in the FIRRTL spec but for which we represent as an exploded
+  /// set of elements in the FIRRTL dialect.
+  UnbundledValuesList unbundledValues;
+
+  /// Provide a symbol table scope that automatically pops all the entries off
+  /// the symbol table when the scope is exited.
+  struct ContextScope {
+    friend class FIRRTLModuleContext;
+    ContextScope(FIRRTLModuleContext &moduleContext, Block *block);
+    ~ContextScope();
+
+  private:
+    void operator=(const ContextScope &) = delete;
+    ContextScope(const ContextScope &) = delete;
+
+    FIRRTLModuleContext &moduleContext;
```

Comment:
Aren't some of these redundant for any code which manages to get a handle to this?